### ISSUE TITLE
[UX] Add help text on "uninstall modules" page /admin/modules/uninstall that modules need to be disabled first

### ISSUE THREAD

**Issue by irinaz** [backdrop/backdrop-issues#6309]
## Description of the need

On the page /admin/modules/uninstall modules are not listed.  In order to see modules there they need to be disabled first.  

## Proposed solution

Add help text explaining how to disable modules.

**Comment by bugfolder** [comment]
I certainly found the distinction between "disable" and "uninstall" a bit non-intuitive when I was starting out in D*, so I agree that some words of explanation here would be useful.

How about this to start the discussion:

"Any modules listed here have been disabled (so that they are no longer operational, but their data still remains). Check the box for any modules to be completely uninstalled, which will entirely remove their settings and/or associated data. You can disable other modules from the [List Modules](#) tab above."

**Comment by irinaz** [comment]
@bugfolder , this would work great! What needs to happen to get this into core?

**Comment by stpaultim** [comment]
@irinaz This kind of change can get into any bug fix release. 

First step is to get a PR with this language and ask for feedback. Feedback is much more likely, once you have a PR. If someone actively monitors and pushes this issue forward, it could get into core pretty quickly. 

But, usually there is some discussion about the specific wording. Bringing this issue to a design/UX meeting could speed up the process. We have a design/UX meeting this week at noon (Pacific).

**Comment by irinaz** [comment]
@stpaultim , @bugfolder , I do not know where this code needs be changed :) If you point me in the right direction I will be happy to open PR.  Thank you in advance!

**Comment by stpaultim** [comment]
@irinaz I would have to look for a similar use case somewhere else and then copy the code. I'll try to look later today, unless @bugfolder is able to help first. 

I think the text can be shortened a bit, which is usually something that comes up. Here is my quick stab at a shorter version that focuses on the important point.s

"Modules listed here have been disabled (they are no longer operational, but their data still remains). Unstalling them will remove their settings and/or any associated data. Modules are disabled from the [List Modules](https://github.com/backdrop/backdrop-issues/issues/6309#) tab above. "

I don't think we need to explain checking the box. That is a pretty common process used in many places. I'm not sure if we need to explain how to 'disable' modules, but I can see how that might be helpful on this page so I included a slightly shorter version of that in my text.

**Comment by klonos** [comment]
Thanks for opening this issue @irinaz and thank you for starting us on a wording suggestion @bugfolder and @stpaultim  👍🏼 🙏🏼 

> I don't think we need to explain checking the box.

I agree 👍🏼 

My suggestion is to try to distinguish disable/uninstall by italicising the words, and by adding separate bullet points for each:
> Any modules listed here have been _disabled_, but not fully _uninstalled_ yet.
> - _Disabling_ module from the "List modules" tab above stops them from being operational, but retains their settings and associated data in case you decide to enable them again in the future.
> - _Uninstalling_ modules, which is what this page allows you to do, will remove their associated data and settings.

**Comment by irinaz** [comment]
@klonos , do you know where PR should go? It should be in core, right?

**Comment by bugfolder** [comment]
> I do not know where this code needs be changed :) If you point me in the right direction I will be happy to open PR.

A search for the page's menu item definition, `$items['admin/modules/uninstall']`, turns up the full item definition:

```
  $items['admin/modules/uninstall'] = array(
    'title' => 'Uninstall modules',
    'page arguments' => array('system_modules_uninstall'),
    'access arguments' => array('administer modules'),
    'type' => MENU_LOCAL_TASK,
    'file' => 'system.admin.inc',
    'weight' => 20,
  );
```

which tells us we can find the function that creates the page by searching for `function system_modules_uninstall(` in file `sysem.admin.inc`.

That function builds a form, so we can add form elements as markup. And since there's a conditional as to whether there are modules to uninstall or not, we can even have different help text for when there are no modules or some.

**Comment by irinaz** [comment]
I would prefer to take the easiest path.  It is OK to have help text in case there are no modules to uninstall for those who are using this feature first time. 

@bugfolder , does this code look correctly with markup?  if yes, I will create fork of core, make change and submit PR

$items['admin/modules/uninstall'] = array(
    'title' => 'Uninstall modules',
    'page arguments' => array('system_modules_uninstall'),
    'access arguments' => array('administer modules'),
    'type' => MENU_LOCAL_TASK,
    'file' => 'system.admin.inc',
    'weight' => 20,
    '#markup' => t('Modules have to be disabled in tab List Modules before they can be uninstalled. Uninstalling module will entirely remove their settings and/or associated data.'),
  );

**Comment by kiamlaluno** [comment]
Route definitions in `hook_menu()` implementations do not use `#markup`. That must be added to the page render array.

**Comment by irinaz** [comment]
@kiamlaluno , thank you so much for response! I am not sure how / where to correctly add this text to page render array. I am more of a site builder and therefore big fan of Backdrop :)

**Comment by kiamlaluno** [comment]
The code needs to be added to [`system_modules_uninstall()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.admin.inc/function/system_modules_uninstall/1).

The code in that function builds the form when there are modules to uninstall. It should show a message in the case there are no module to uninstall.

**Comment by irinaz** [comment]
@kiamlaluno , it would be in [core/modules/system/system.admin.inc](https://github.com/backdrop/backdrop/blob/affb7c24a2176e30ccc91c6f70193957962f83b8/core/modules/system/system.admin.inc#L1072), correct?

**Comment by klonos** [comment]
> it would be in [core/modules/system/system.admin.inc](https://github.com/backdrop/backdrop/blob/affb7c24a2176e30ccc91c6f70193957962f83b8/core/modules/system/system.admin.inc#L1072), correct?

That is correct @irinaz 👍🏼

**Comment by kiamlaluno** [comment]
In this case, the added code could be similar to the following one.

```php
$form['help_text'] = array(
  '#theme' => 'help',
  '#markup' => t('Modules must be first disabled from …'),
);
// Only build the rest of the form if there are any modules available to
// uninstall.
if (!empty($disabled_modules)) {
  // Omissis
}
```

The comment in that code is the one already present in [`system_modules_uninstall()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.admin.inc/function/system_modules_uninstall/1).

**Comment by irinaz** [comment]
I was working on another site and noticed that by default there is a message "No modules are available to uninstall."  We should modify this message to add same text as in the case when there are some modules that can be uninstalled. 

t('Modules have to be disabled in tab List Modules before they can be uninstalled. Uninstalling module will entirely remove their settings and/or associated data.'),

**Comment by stpaultim** [comment]
@irinaz I will continue to work on this PR. The text is akward at the moment, I'll try to fix that a bit and improve the formatting a bit.

**Comment by stpaultim** [comment]
Ready for feedback. See sandbox attached to PR. 

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/2bc95df0-755b-4675-8132-0f9e32f43bb7)

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/022b63ed-b47c-4639-b7e3-71475c97e9b3)

**Comment by irinaz** [comment]
This looks great, thank you!!

**Comment by stpaultim** [comment]
Should we add one more sentence that says something to the effect. 

"Uninstalled modules remain in the modules folder and can be enabled again in the future."

**Comment by stpaultim** [comment]
Added "the" to "in the List Modules tab." @docwilmot 

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/360d1f25-9f7e-486a-8fff-abe4fe8115ce)

**Comment by klonos** [comment]
> I think the same text works in both situations. Or?

Yes 👍🏼 ...even if some modules exist in the uninstall list/table, one might come here expecting to see another module that has not yet been disabled, and wonder. In that case, the `Modules have to be disabled in the List modules tab before they can be uninstalled` bit is useful and provides clarity.

> Should we add one more sentence that says something to the effect.

You could combine that with the last sentence:

> ... Uninstalling a module will entirely remove its settings and/or any associate data, while leaving the module itself available in case it needs to be enabled again in the future.

I believe that adding information that explains how things work should not be sacrificed over brevity. Yes, we should try to keep help text as short as possible, but not to the point where it is not accurate or missing important details. So I would also add something to explain that if a module needs to be removed from the site forever, then it will need to be manually removed from the codebase.

**Comment by quicksketch** [comment]
The PR as-is looks pretty good to me.

**Comment by klonos** [comment]
It does need some work:
- The m needs to be lowercase in `"List modules" tab`. Although the primary tabs are upper-cased via CSS, it's still `List modules`:
  ![image](https://github.com/backdrop/backdrop-issues/assets/2423362/460382a9-046b-42e9-b811-a9d5cfe78d92)
  Besides, the only cases were we capitalize the letters of words is for module/theme/layout names.
- Would be nice to explain that the module itself (the code) is not removed from the site, why that is (in order to be able to enable it again in the future), and what needs to be done for it to actually be removed.
- I would still like us to clarify the difference between disabling vs. unistalling.

**Comment by olafgrabienski** [comment]
> and what needs to be done for it to actually be removed.

@stpaultim Looking at the screenshot, I guess the part about removal of a module is still missing.

**Comment by albanycomputers** [comment]
I agree with @klonos we need to inform people that the files have not been removed.

I also think the test should be something like...

"The modules listed below have been disabled... To remove the module's settings and data you need to uninstall them, this will not remove the module files they need to be removed manually. "

Do we need to let people know where or how to disable modules?  i.e. if they are in the "Uninstall modules" then they should be aware of how and where modules are installed and how to check the boxes... sometimes we oversimplify things with too much information and other times we leave out or make things harder to comprehend.

**Comment by indigoxela** [comment]
Looks great. A little wordy, but I'm also fine with it.

**Comment by klonos** [comment]
> In the UK we have a saying... "Teaching your grandmother to suck eggs"

In Greece we have a similar (but much more rude/explicit) saying, involving grandparents and teaching them about procreation 😅 ...same thing though.

> The modules listed below have been disabled

The help text we are introducing here will be shown even when that list doesn't have any modules. In that case it wouldn't make sense to say that, and if we wanted to say that, then we'd have to add logic to be conditionally prepending that bit to the help text. Not worth it I believe.

> ... they should be aware of how and where modules are installed and how to check the boxes ...

I agree about the checkboxes bit, but not about where modules are enabled and disabled being obvious/apparent. The "Install new modules" and "Uninstall modules" tabs both have self-explanatory labels, but "List modules" doesn't convey what happens in that tab. ...and we don't want to make that label too wordy either ("List and enable/disable modules"). So yeah, I think we should keep the `Modules have to be disabled in the List modules tab ...` bit.

**Comment by cellear** [comment]
I think it's an elegant and useful addition, and one that's long overdue.

**Comment by argiepiano** [comment]
I like the wording!

**Comment by bugfolder** [comment]
LGTM.

**Comment by albanycomputers** [comment]
> > The modules listed below have been disabled
> 
"Any modules listed below have been disabled" :-)


> I agree about the checkboxes bit, but not about where modules are enabled and disabled being obvious/apparent. The "Install new modules" and "Uninstall modules" tabs both have self-explanatory labels, but "List modules" doesn't convey what happens in that tab. ...and we don't want to make that label too wordy either ("List and enable/disable modules"). So yeah, I think we should keep the `Modules have to be disabled in the List modules tab ...` bit.

Change the "List modules" and add "Enabled" at the top of the column for the check box.

**Comment by albanycomputers** [comment]
"Any modules listed below are disabled"

**Comment by kiamlaluno** [comment]
If the message is shown also when the list is empty, I would not show *The modules listed below …*. I would rather use: 

> Modules must be first disabled in the [List modules](#) tab to be uninstalled. Uninstalling modules will permanently remove their settings and/or data; the modules themselves are still available in the case they need to be re-enabled.

**Comment by bugfolder** [comment]
> If the message is shown also when the list is empty, I would not show The modules listed below...

The current form logic already includes a test of whether the list is empty, so it would be straightforward and appropriate to have different messages based on whether or not there are listed modules.

Also a bit of logic efficiency: no need to add the `'help'` item to `$form` till after the decision whether to return the `$confirm_form`.

Suggested changes in [this PR](https://github.com/backdrop/backdrop/pull/4591). (I'm happy to close the PR if you'd rather keep changes in the original.)

**Comment by stpaultim** [comment]
Here are screenshots for the most recent PR by @bugfolder. (Thanks @bugfolder for providing this.)

This is how it currently looks with no disabled modules available to uninstalled.
![image](https://github.com/backdrop/backdrop-issues/assets/3144571/d0588f93-2505-4cbc-92b8-149e3fbecfd4)

Here is how it looks if there are modules to be installed.
![image](https://github.com/backdrop/backdrop-issues/assets/3144571/3f07db1a-14db-422c-9184-49e03a498e9e)

I don't think that this logic was required. I think it read fine without the extra details. But, I don't object to this additional text. 

> I guess the part about removal of a module is still missing.

There is no text specifically mentioning that a module files must be removed. But, there is text that at least makes it explicit that this features intentionally does not remove files. I would assume that anyone with the ability to remove the files, can figure the rest out. What was confusing, was that some people expected that files would be removed and they were not. This text makes it explicit that that files are not intended to be removed.

In general, I think we have worked very had to limit help text in the UI, because help text can be a crutch to mask poor UI. It's our preference to make the feature more intuitive to use, rather than simply providing text to help people better understand a bad UI. We have generally worked to keep help text as brief and easy to understand as possible. 

Having said that, I'm not sure how to make this UI better without adding a little help text, so I'm pretty happy with what we have (although, I'm nervous about making it any longer).

**Comment by olafgrabienski** [comment]
>> I guess the part about removal of a module is still missing.

> There is no text specifically mentioning that a module files must be removed. But, there is text that at least makes it explicit that this features intentionally does not remove files.

While module files don't *have to* be removed, I still think it's important to mention how they *can* be removed. This might be one sentence more, make the new text, which is already a great improvement, much clearer, in my opinion.

**Comment by bugfolder** [comment]
> In general, I think we have worked very had to limit help text in the UI, because help text can be a crutch to mask poor UI.

Well, we already have the UI (this page, this tab, this table). Once the UI is defined, help text can, um, help. And from the discussion above, it sounds like having some help text here is indeed helpful.

If the help text gets too long, we could wrap it in an expandable `details` element as we do with "How layouts work" on admin/structure/layouts. But I don't think we're there yet on this page.

(And if one thinks that this page is "poor UI" that the help text is masking, I would suggest that would be better addressed by filing a seperate issue with a proposal for the new and improved UI.)

> I still think it's important to mention how they can be removed. This might be one sentence more...

@olafgrabienski, could you suggest some wording for that new sentence?

**Comment by olafgrabienski** [comment]
Sure, I can try!

Current PR:

> Modules must be disabled in the [List modules](https://pr4591-isfli0tjcgvv3pvuerw4ulducy0qzw9u.tugboatqa.com/admin/modules) tab before they can be uninstalled. Uninstalling a module will permanently remove its settings and/or associated data, while leaving the module files in place in case it needs to be enabled again in the future.

One more sentence:

> Modules must be disabled in the [List modules](https://pr4591-isfli0tjcgvv3pvuerw4ulducy0qzw9u.tugboatqa.com/admin/modules) tab before they can be uninstalled. Uninstalling a module will permanently remove its settings and/or associated data, while leaving the module files in place in case it needs to be enabled again in the future. To remove a module permanently, delete its files from the `modules` folder of your Backdrop installation.

**Comment by irinaz** [comment]
What is next step in finalizing this PR? Thank you!

**Comment by stpaultim** [comment]
@bugfolder I'm fine with the changes suggested by @olafgrabienski. I think its getting long enough that the "Details" element might be warranted. BUT, I'm also happy to leave it is as it is now to keep things moving.

Having added the text suggested by @olafgrabienski, I think it would be ok to shorten  "while leaving the module files in place in case it needs to be enabled again in the future" to "but leave the module files in place." 

Something like this:

_"Modules must be disabled in the [List modules] tab before they can be uninstalled. Uninstalling a module will permanently remove its settings and/or associated data, but leave the module files in place. To remove a module permanently, delete its files from the modules folder of your Backdrop installation."_

But, I'm also fine with the current wording. I'll close my PR, since the new one has momentum.

**Comment by stpaultim** [comment]
Looks excellent.

**Comment by olafgrabienski** [comment]
Thanks for the updates, I like the current version as well!

One thing I was wondering about: What's the best / common way to write folder names? I tend to put paths in `<code>` tags, and IIRC I've also seen `<em>` for paths. In the PR, the path is indicated by a starting slash. Any preferences?

- delete its files from the `modules` folder
- delete its files from the *modules* folder
- delete its files from the /modules folder

**Comment by klonos** [comment]
@bugfolder I've added a suggestion re the help text variables being used. Other than that, this is RTBC 👍🏼

**Comment by bugfolder** [comment]
@klonos, good idea; that also allows to eliminate the duplicated `$form['help']`. PR updated.

**Comment by quicksketch** [comment]
Thanks folks! I like where this ended up. I merged https://github.com/backdrop/backdrop/pull/4591 into 1.x and 1.26.x. Thanks everyone for their feedback and especially @bugfolder for updating the PR nearly a dozen times.

**Comment by irinaz** [comment]
thank you, mucho gracias!
